### PR TITLE
Purge MINIFIER_SPAWNED

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90089
* #90039

Purge MINIFIER_SPAWNED

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/optimizations/backends.py [hunk 1]
```diff
@@ -750,7 +750,6 @@ def torchxla_trace_once(subgraph):
     compiled_graph = None
 
     def fwd(*args):
-        nonlocal subgraph
```
Comment (ezyang) [inline torch/_dynamo/optimizations/backends.py:753]: what's going on here?

File: torch/_dynamo/debug_utils.py [hunk 2]
```diff
@@ -569,7 +569,7 @@ def same_two_models(gm, opt_gm, example_inputs, only_fwd=False):
                 " Skipping this graph."
             )
         )
-        return True
+        raise
```
Comment (ezyang) [inline torch/_dynamo/debug_utils.py:572]: ??

Comment (voznesenskym) [reply torch/_dynamo/debug_utils.py:572]: We need this as it moved to runtime

File: torch/_dynamo/optimizations/backends.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/optimizations/backends.py:753]: Bad commit

File: torch/_dynamo/debug_utils.py [hunk 3]
```diff
@@ -962,33 +958,43 @@ def dynamo_accuracy_minifier_backend(gm, example_inputs, compiler_name):
...
+        ), f"Unknown compiler name {compiler_name} provided."
         compiler_fn = BACKENDS[compiler_name]
 
-    # Set the eval mode to remove randomness.
-    gm.eval()
+    # (2) Compile the graph
+    compiled_gm = compiler_fn(copy.deepcopy(gm), clone_inputs(example_inputs))
 
-    # Check Accuracy
-    if backend_accuracy_fails(gm, example_inputs, compiler_fn):
```
Comment (ezyang) [inline torch/_dynamo/debug_utils.py:976]: What happened to this call?

Comment (voznesenskym) [reply torch/_dynamo/debug_utils.py:976]: `if same_two_models` this call used to wrap it, we dont need the wrapper.

`pytest test/dynamo/test_minifier.py test/inductor/test_minifier.py` tests this :)

Comment (ezyang) [reply torch/_dynamo/debug_utils.py:976]: So, the `backend_accuracy_fails` is inlined now. Does this mean there are two copies of the code? Should we delete the old one?

File: torch/_dynamo/debug_utils.py [hunk 4]
```diff
@@ -962,33 +958,43 @@ def dynamo_accuracy_minifier_backend(gm, example_inputs, compiler_name):
...
-            example_inputs,
-            module_fails=fails_fn,
-            dump_state=dump_state_fn,
-        )
-    else:
-        log.error("Input graph does not fail accuracy testing")
-    return gm
+    def backend_accuracy_eval_fwd(*args):
+        # Set the eval mode to remove randomness.
+        gm.eval()
```
Comment (ezyang) [inline torch/_dynamo/debug_utils.py:977]: This looks like an unnecessary behavior change; previously gm was changed to eval() mode at compile time, now it is reset to eval time every runtime.

...arguably we shouldn't be changing the model to eval mode at all, doesn't this mean running with accuracy minifier changes the results of your model...

File: torch/_dynamo/debug_utils.py [hunk 5]
```diff
@@ -962,33 +958,43 @@ def dynamo_accuracy_minifier_backend(gm, example_inputs, compiler_name):
...
+
+        def module_fails_accuracy(gm, args):
+            return not same_two_models(gm, compiled_gm, args)
+
+        # (4) Check Accuracy
+        if not same_two_models(gm, compiled_gm, args):
```
Comment (ezyang) [inline torch/_dynamo/debug_utils.py:983]: Use `module_fails_accuracy`?

File: torch/_dynamo/debug_utils.py [hunk 6]
```diff
@@ -962,33 +958,43 @@ def dynamo_accuracy_minifier_backend(gm, example_inputs, compiler_name):
...
+            log.warning("Accuracy failed for the TorchDyanmo produced graph")
+            dump_state_fn = functools.partial(
+                dump_backend_state, compiler_name=compiler_name, check_accuracy=True
+            )
+
+            dump_state_fn(fx.GraphModule(gm, copy.deepcopy(gm.graph)), args)
+            minifier(
+                gm,
+                example_inputs,
+                module_fails=module_fails_accuracy,
```
Comment (ezyang) [inline torch/_dynamo/debug_utils.py:993]: This looks questionable. The original fails_fn recompiles the passed in gm. However, your fails_fn here simply reuses the original compiled_gm as is. This seems wrong: I'd expect the minifier to be modifying the gm passed in and expecting it to recompile. Without recompiling it I find it hard to see how the minifier could do the correct thing here.

**Review by ezyang — CHANGES_REQUESTED** [review]
The new code looks buggy

File: torch/_dynamo/debug_utils.py (hunk 6 above)
Comment (voznesenskym) [reply torch/_dynamo/debug_utils.py:993]: Good point, I bet it minifies but not the same way, as the backend is taken out. Let me revisit this.

File: torch/_dynamo/debug_utils.py (hunk 5 above)
Comment (voznesenskym) [reply torch/_dynamo/debug_utils.py:983]: Sure :)

File: torch/_dynamo/debug_utils.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dynamo/debug_utils.py:976]: We can collapse them together with a little modification, its used elsewhere

File: torch/_dynamo/debug_utils.py (hunk 4 above)
Comment (voznesenskym) [reply torch/_dynamo/debug_utils.py]: Good point, let's go back to the prior way where we ensure we compile.

File: torch/_dynamo/debug_utils.py (hunk 2 above)
Comment (ezyang) [reply torch/_dynamo/debug_utils.py:572]: OK, well, if you want to change the behavior here, you need to at least update the warning above.

But the original intent of squelching the error still seems valid; if you get stuck on an unrelated exception, you're never going to get to the actual failure that you wanted to minify. So I don't really understand your stated justification...

File: torch/_dynamo/debug_utils.py [hunk 7]
```diff
@@ -757,14 +762,13 @@ def backend_accuracy_fails(gm, example_inputs, compiler_fn, only_fwd=False):
         )
         return False
 
-    return not same_two_models(gm, compiled_gm, example_inputs, only_fwd)
+    if real_args is None:
+        real_args = example_inputs
```
Comment (ezyang) [inline torch/_dynamo/debug_utils.py:766]: Is example inputs expected to be non-fake if real args is None?

File: torch/_dynamo/debug_utils.py [hunk 8]
```diff
@@ -962,33 +966,51 @@ def dynamo_accuracy_minifier_backend(gm, example_inputs, compiler_name):
...
-            example_inputs,
-            module_fails=fails_fn,
-            dump_state=dump_state_fn,
-        )
-    else:
-        log.error("Input graph does not fail accuracy testing")
-    return gm
+            # (4) Bind the current compiler_fn and real_args to accuracy_eval
+            # We need to do this in order to ensure that (a) we correctly recompile on every
+            # minification, and (b) that the args captured here, in the runtime, are passed to eval
```
Comment (ezyang) [inline torch/_dynamo/debug_utils.py:993]: This comment is backwards. To "correctly recompile on every minification", the bug in https://github.com/pytorch/pytorch/pull/90089/files/ae9f856428538468b5da8825e9d796790b010db6#r1038871303 was that you didn't pass `backend_accuracy_fails` to the minifier at all. Binding compiler_fn/real_args has nothing to do with ensuring that we actually recompile; you must bind them for the simpler reason that the minifier isn't going to pass those arguments directly lol

**Review by ezyang — APPROVED** [review]
I'm still very skeptical about the change in `same_two_models`. There are also comments I'd like to see addressed.

File: torch/_dynamo/debug_utils.py (hunk 8 above)
Comment (voznesenskym) [reply torch/_dynamo/debug_utils.py:993]: > This comment is backwards. To "correctly recompile on every minification", the bug in https://github.com/pytorch/pytorch/pull/90089/files/ae9f856428538468b5da8825e9d796790b010db6#r1038871303 was that you didn't pass `backend_accuracy_fails` to the minifier at all. Binding compiler_fn/real_args has nothing to do with ensuring that we actually recompile; you must bind them for the simpler reason that the minifier isn't going to pass those arguments directly lol

I was passing a different function, identical to this except for recompilation. So now I'm confused by what you mean.

Comment (ezyang) [reply torch/_dynamo/debug_utils.py:993]: The comment says "Bind the current compiler_fn and real_args to accuracy_eval; we need to do this". I parse this as "we need to [Bind the current compiler_fn and real_args]". Let's suppose I removed this binding. Then I would expect this code to fail with "not enough parameters passed".

**Pull request closed without merging**